Simplification of grammar that encodes mutations with the same productions as reactions

Wonseo-C opened the pull request:
This PR:
* Updates the Lingua Franca grammar to delete the Mutation class, and merge it with the Reaction class. (They have the similar properties `Trigger`, `Source`, `Effect`, `Code`)
* Updates the TypeScript code generator to uses a `WritablePort` for mutations

hokeun commented:
@lhstrh, I talked to @Wonseo-C and we found that [`Reaction`](https://github.com/lf-lang/lingua-franca/blob/master/org.lflang/src/org/lflang/LinguaFranca.xtext#L197) and [`Mutation`](https://github.com/lf-lang/lingua-franca/blob/master/org.lflang/src/org/lflang/LinguaFranca.xtext#L219) have a lot in common. Do you know if we can define a parent of these two objects in [`LinguaFranca.xtext`](https://github.com/lf-lang/lingua-franca/blob/master/org.lflang/src/org/lflang/LinguaFranca.xtext)? If we have a sort of "parent class" of [`Reaction`](https://github.com/lf-lang/lingua-franca/blob/master/org.lflang/src/org/lflang/LinguaFranca.xtext#L197) and [`Mutation`](https://github.com/lf-lang/lingua-franca/blob/master/org.lflang/src/org/lflang/LinguaFranca.xtext#L219), we will be able to write code a lot cleaner without repeating the same code. :) For example, we can define common functions used by procedures creating reactions and mutations.

lhstrh commented on org.lflang/src/org/lflang/generator/ts/TSReactionGenerator.kt:
```diff
@@ -103,43 +107,66 @@ class TSReactionGenerator(
             """
             |
-            |this.addReaction(
+            |this.add${if (reaction.isMutation()) "mutation" else "reaction"}(
```
```suggestion
            |this.add${if (reaction.isMutation()) "Mutation" else "Reaction"}(
```

lhstrh commented on org.lflang/src/org/lflang/generator/ts/TSReactionGenerator.kt:
```diff
@@ -180,6 +207,21 @@ class TSReactionGenerator(
         }
     }
 
+    private fun generateMutationSignatureElementForPortEffect(effect: VarRef): String {
```
Comments missing

Wonseo-C replied:
I wrote the comments for this function in https://github.com/lf-lang/lingua-franca/pull/1318/commits/dbec99308314d3e9cb63d9f47ace6c83530d1bfc

lhstrh commented on org.lflang/src/org/lflang/scoping/LFScopeProviderImpl.java:
```diff
@@ -257,7 +259,7 @@ protected IScope getScopeForVarRef(VarRef variable, EReference reference) {
         if (variable.eContainer() instanceof Deadline) {
             return RefType.DEADLINE;
-        } else if (variable.eContainer() instanceof Reaction) {
+        } else if (variable.eContainer().getClass() == ReactionImpl.class) {
```
```suggestion
        } else if (variable.eContainer() instanceof Reaction) {
```

lhstrh commented on org.lflang/src/org/lflang/scoping/LFScopeProviderImpl.java:
```diff
@@ -266,7 +268,7 @@ private RefType getRefType(VarRef variable) {
                 return RefType.EFFECT;
             }
-        } else if (variable.eContainer() instanceof Connection) {
+        } else if (variable.eContainer().getClass() == ConnectionImpl.class) {
```
```suggestion
        } else if (variable.eContainer() instanceof Connection) {
```

Wonseo-C commented:
* I deleted the unnecessary code lines (https://github.com/lf-lang/lingua-franca/pull/1318/commits/fbec70a2776aa3fa7bbd7a4bef12533a689de94e, https://github.com/lf-lang/lingua-franca/pull/1318/commits/4d49227638e48405bb262cebf498baaf6081f0b8)
* Combined the function of generating Mutation and Reaction (https://github.com/lf-lang/lingua-franca/pull/1318/commits/e444eb44f94360cbf4b78e90570d2f0d0ce0ae84)
* Added comment for `generateMutationSignatureElementForPortEffect` (https://github.com/lf-lang/lingua-franca/pull/1318/commits/dbec99308314d3e9cb63d9f47ace6c83530d1bfc)

[Sieve.lf](https://github.com/lf-lang/benchmarks-lingua-franca/blob/main/TS/Savina/src/parallelism/Sieve.lf), the simple example file using the `mutation`, is built well. @lhstrh

lhstrh commented on org.lflang/src/org/lflang/generator/ts/TSImportPreambleGenerator.kt:
```diff
@@ -55,7 +55,7 @@ class TSImportPreambleGenerator(
             |import {Bank as __Bank} from './core/bank'
             |import {FederatedApp as __FederatedApp} from './core/federation'
-            |import {InPort as __InPort, OutPort as __OutPort, Port as __Port} from './core/port'
+            |import {InPort as __InPort, OutPort as __OutPort, Port as __Port, WritablePort} from './core/port'
```
```suggestion
            |import {InPort as __InPort, OutPort as __OutPort, Port as __Port, WritablePort as __WritablePort} from './core/port'
```

lhstrh commented on org.lflang/src/org/lflang/generator/ts/TSReactionGenerator.kt:
```diff
@@ -1,5 +1,6 @@
 package org.lflang.generator.ts
 
+import org.eclipse.emf.ecore.EObject
```
```suggestion
```

lhstrh commented on org.lflang/src/org/lflang/generator/ts/TSReactionGenerator.kt:
```diff
@@ -103,43 +107,66 @@ class TSReactionGenerator(
-                    BuiltinTrigger.SHUTDOWN -> reactionTriggers.add("this.shutdown")
-                    else -> {}
+
+        if (reaction.isMutation()) {
```
Why are these cases handled separately? They look the same to me...

Wonseo-C replied:
Yes, I deleted the duplicate code. https://github.com/lf-lang/lingua-franca/pull/1318/commits/0d49d6e37ef948cd22628394e533ffab613e524b

lhstrh commented on org.lflang/src/org/lflang/generator/ts/TSReactionGenerator.kt:
```diff
@@ -103,43 +107,66 @@ class TSReactionGenerator(
     ): String {
         // Assemble reaction triggers
+        var reactCode = ""
+        var reactDeadline = "}"
```
Why is the variable initialized with a closing curly bracket?

Wonseo-C replied:
I defined the `reactDeadline` as `"}"`, because when the deadline is null, then the deadline should be empty.
So, at [this line](https://github.com/lf-lang/lingua-franca/pull/1318/files#diff-5fdb14bc11c18f34a23a51db49bd847b0bc8b386cc889dff06cb825d2ab20cf5R119), I replaced the value of `reactDeadline` if the reaction's deadline isn't null.

lhstrh replied:
I think the old line `139` is actually much more self-explanatory. I.e.: 
```
${" |    "..if (reaction.deadline != null) generateDeadlineHandler(reaction, reactPrologue, reactEpilogue, reactSignature) else "}"}
```

Wonseo-C replied:
Thank you for your nice code suggestion! I reflected your comment in [f327fee](https://github.com/lf-lang/lingua-franca/commit/f327feef4c499641d83833eb1fd3591f20c35834).

lhstrh commented on org.lflang/src/org/lflang/generator/ts/TSReactionGenerator.kt:
```diff
@@ -333,8 +379,13 @@ class TSReactionGenerator(
             } else if (effect.variable is Port){
-                reactSignatureElement += ": ${generateReactionSignatureElementForPortEffect(effect)}"
-                reactEpilogue.add(generateReactionEpilogueForPortEffect(effect))
+                if (reaction.isMutation()) {
```
Duplicate code...

Wonseo-C replied:
I deleted the duplicate code. https://github.com/lf-lang/lingua-franca/pull/1318/commits/d58c06655ed025e9537ca8b9dbf5c5af04316984

lhstrh commented on org.lflang/src/org/lflang/generator/ts/TSReactionGenerator.kt:
```diff
@@ -449,6 +500,7 @@ class TSReactionGenerator(
         // If the app is federated, only generate
         // reactions that are contained by that federate
         val generatedReactions: List<Reaction>
+        val generatedMutations: List<Reaction>
```
Why is the variable called `generatedMutations`? I think just `mutations` is more accurate.

Wonseo-C replied:
I changed the variable name! https://github.com/lf-lang/lingua-franca/commit/b349e6010a667132c40f78a71128e268fecd00dc

lhstrh commented on org.lflang/src/org/lflang/generator/ts/TSReactionGenerator.kt:
```diff
@@ -464,11 +516,20 @@ class TSReactionGenerator(
+        if (generatedMutations.isNotEmpty()) {
+            // Write the mutation itself
+            for (mutation in generatedMutations) {
+                reactionCodes.add(generateSingleReaction(reactor, mutation))
```
Does this method account for the fact that the given reaction is a mutation? If so, then why are we handling mutations and reactions separately at all?

lhstrh replied:
My question is: why do we gather mutations separately from reactions if for both we end up calling `generateSingleReaction`, anyway?

Wonseo-C replied:
Yes, you're right. I was mistaken that the `Reaction` and `Mutation`  were separated.
I deleted this part. https://github.com/lf-lang/lingua-franca/pull/1318/commits/61bb0286a4433f3d8c16efa3ed4ba017c7d307b0

lhstrh commented on org.lflang/src/org/lflang/generator/ts/TSReactionGenerator.kt:
```diff
@@ -103,11 +107,20 @@ class TSReactionGenerator(
+        var reactDeadline = "}"
         val reactionTriggers = StringJoiner(",\n")
+
+        if (!reaction.isMutation()) {
```
Can mutations not have deadlines? I think this can all go away and we can just reinstate the original line `139`.

Wonseo-C replied:
You're right. I deleted this part. https://github.com/lf-lang/lingua-franca/pull/1318/commits/f327feef4c499641d83833eb1fd3591f20c35834

lhstrh commented on org.lflang/src/org/lflang/generator/ts/TSReactionGenerator.kt:
```diff
@@ -119,27 +132,29 @@ class TSReactionGenerator(
                 }
             }
         }
+        reactCode = reaction.code.toText()
```
This variable also seems unnecessary...

Wonseo-C replied:
Thank you. I deleted the unnecessary variable! https://github.com/lf-lang/lingua-franca/pull/1318/commits/f327feef4c499641d83833eb1fd3591f20c35834

lhstrh commented on org.lflang/src/org/lflang/generator/ts/TSReactionGenerator.kt:
```diff
@@ -119,27 +124,28 @@ class TSReactionGenerator(
             |    function ($reactSignature) {
             |        // =============== START react prologue
-        ${" |        "..reactPrologue}
+            ${" |        "..reactPrologue}
```
The indentation here is weird. Same on line 139, 142, 145, etc.

Wonseo-C replied:
https://github.com/lf-lang/lingua-franca/pull/1318/commits/e07dc3972dac5adfd4a7e864607f6f746cb6f049

lhstrh commented on org.lflang/src/org/lflang/generator/ts/TSReactionGenerator.kt:
```diff
@@ -180,6 +186,25 @@ class TSReactionGenerator(
+     * Generate the mutation's effect type as "WritablePort,"
+     * which differs from the reaction's effect type.
+     */
+    private fun generateMutationSignatureElementForPortEffect(effect: VarRef): String {
```
I realize that you had to make this distinction because we set the data type up this way in reactor-ts, but it wasn't a very well-though-out decision on my part to implement it that way. It doesn't hurt reactions to be able to inspect the port, so I suggest that instead augment `ReadWrite` with `getPort()` and `MultiReadWrite` with `getPorts` and remove the distinction between reactions and mutations. WDYT?

Wonseo-C replied:
@lhstrh I think it is an excellent idea!
But I have a question; is the `WritablePort` not necessary? If `WritablePort` is deleted, then there will be many problems with implementation. The reason is the WritablePort is class, and the ReadWrite is type.

Wonseo-C replied:
I tried to fix it, but there are some problems.
If we add the `getPort()` in the `ReadWrite`, we should get the `IOPort`. So, the `ReadWrite`'s type should be `Present`. It caused a type error. (The `ReadWrite` is no class)
I think that we can use the `asWritable()`, which is the `IOPort`'s function of returning the `writablePort`.

https://github.com/Wonseo-C/reactor-ts/tree/getport : This is the branch where I worked.

Wonseo-C replied:
@lhstrh :
By discussing with @hokeun, we guessed your purpose of this review is to make the simple code. And we decided to implement it like this -> https://github.com/lf-lang/lingua-franca/pull/1318/commits/74ad4094bc837200e4667eb3864f78b8b9aa326b.
We thought your suggestion could cause another problem because of the type error (detail explained above), so I just modified the `generateReactionSignatureElementForPortEffect` to get different strings if it is a mutation or not.

lhstrh reviewed (approved):
Looks good! Thanks for these fixes!

hokeun commented:
Great work @Wonseo-C ! Thanks for the review @lhstrh !
